proposals/TP3-xds-error-propagation.md

```diff
@@ -0,0 +1,122 @@
+
+The xDS protocol does have a way for xDS clients to NACK responses back from the xDS Management server. The NACK response also contains an `error_detail` field which the Management Server can use to extract further information about the rejection. But this NACK’ing mechanism is restricted to the client i.e. there is no way for Management Server to actually convey a notification(Ex: unavailability, permission issues etc) regarding some or all the resource requests that are being subscribed by the client. This eventually leads to the client timeouts which, although it conveys an error back to the application, it misses the actual context for the issue. In most cases the xDS Management Servers might not even be accessible for the client applications to debug these issues without escalation. 
+
+### Related Proposals:
+
+N/A
+
+## Proposal
+
+The proposal introduces two new fields to the DiscoveryResponse proto. The first, `resource_errors`, will detail Resource-specific errors and be included in both SotW and Incremental xDS protocols. The second, `removed_resources`, will specifically identify resources that no longer exist, addressing a current gap in the SotW protocol's signaling capabilities. 
+
+```textproto
+// New Proto Message
+// Contains detailed error detail for a single resource
+message ResourceError {
+ ResourceName resource_name = 1;
+
+ google.rpc.Status error_detail = 2;
+}
+
+message DiscoveryResponse {
+
+// The version of the response data.
+string version_info = 1;
+
+….
+…
+
+
+ // The control plane instance that sent the response.
+  config.core.v3.ControlPlane control_plane = 6;
+
+// NEW_FIELD
+// An optional repeated list of error details that the control plane 
+// can provide to notify the client of issues for the resources that 
+// are subscribed.
+// This allows the xDS management server to provide optional 
+// notifications in case of unavailability, permissions errors 
+// without the client having to wait for the `config fetch` timeout.
+repeated ResourceError resource_errors = N;
+
+// NEW FIELD
+// For non-existing resources.
+repeated ResourceName removed_resources = N;
+}
+
+message DeltaDiscoveryResponse {
+  …
+
+  // NEW FIELD
+  repeated ResourceError resource_errors = N;
+}
+```
+
+### Protocol Behavior
+The client must use this additional field to obtain notification for resources the xDS Management server couldn’t procure. The xDS client should cancel any resource timers once this message is received and convey the error message to the application. With the addition of this field, when a client receives an explicit error or does-not-exist indicator from the management server, it should react the same way it would have if its does-not-exist timer fired. 
+
+The xDS Management server is only expected to return the error message once rather than throughout for future responses. The client is expected to remember the error message until either a new error message is returned or the resource is returned. 
```

Comment:
Only _expected_ or is _required not to_?

Reply:
This statement is to clarify that the xDS management server is free to send the error back for future messages too rather than just for the first one. I am not sure if we want to codify this with a stronger statement  i.e. the xDS management server should return the error message *only once*. 

Do you see a benefit by making this a requirement?

Reply:
No I think it's good to not make it "exactly once", just wanted to clarify. thanks

Reply:
I think that if we want to give the control plane the option of not resending every time, then we need to require the client to remember the state.

We should also explicitly state that even for LDS and CDS, where the control plane is required to send every subscribed resource in every response, the control plane does not need to resend the error messages in every response.  The client is required to remember the most recent error state for each resource even for LDS and CDS.

Reply:
Done